### ISSUE TITLE
Search provides wrong results for "Ferraz 4 Madrid"

### ISSUE THREAD

**Issue by biodranik** [organicmaps/organicmaps#7616]
I love Organic Maps, especially the fact that I don't share my location with Google or others.
I have no problems with the navigation, which generally works quite well, despite not having traffic management (I can live without it), but the worst thing is the address search. When I enter an address many times the street does not appear or streets appear that are not the one I am looking for. For example, when searching for "Ferraz 4 Madrid" the "Madrid street" appears in towns that are relatively close or not to my location but I'm searching "Ferraz" street at city of Madrid. How could I do this to narrow my search to the results I really want? (I don't have that problem with Google Maps, for example).

_Originally posted by @kslcsdalsadg in https://github.com/organicmaps/organicmaps/discussions/7044_

**Comment by akashwarrior** [comment]
it appears that search results may be prioritizing relevance based on individual terms rather than considering the complete search phrase. Specifically, when users input queries such as "Ferraz 4 Madrid" or "Ferraz Madrid," the search algorithm seems to give undue weight to the term "Madrid" rather than considering the entirety of the input.

**Comment by kslcsdalsadg** [comment]
> it appears that search results may be prioritizing relevance based on individual terms rather than considering the complete search phrase. Specifically, when users input queries such as "Ferraz 4 Madrid" or "Ferraz Madrid," the search algorithm seems to give undue weight to the term "Madrid" rather than considering the entirety of the input.

Yes, maybe it's as you explain, but the result is that the search cannot be used to query unknown locations, because you may end up in another place miles away...
What I do now to avoid making mistakes is:
1. Search for address on Google Maps using an incognito tab in a web browser
2. Copy the LAT/LON location of the location from Google Maps
3. Find the LAT/LON location in Organic Maps
4. Navigate to the location using Organic Maps
But thus, the main objective of not being tracked by Google is surely lost.

**Comment by muralito** [comment]
I haven't read the code, but it seems that the feature that you call "searching algorithm" is a mix of some kind of NLP and then searching the database, but the NLP implementation seems not robust enough, so OM in several cases fails to understand what the user is looking for.

Does OM uses a NLP library or their own method?

**Comment by muralito** [comment]
In the proposed case, it should be clear to the searching algorithm that the user looks for an address:
city: Madrid
street: Ferraz
housenumber: 4

**Comment by kslcsdalsadg** [comment]
> In the proposed case, it should be clear to the searching algorithm that the user looks for an address: city: Madrid street: Ferraz housenumber: 4

I need to set search text to: "City: Madrid, Street: Fedraz, housenumber: 4" when I have to search to "Fedraz 4, Madrid" ?

**Comment by kslcsdalsadg** [comment]
There's no one difference if I search "Ferraz 4 Madrid", "Ferraz 4, Madrid" or "cify: Madrid, street: Ferraz 4"  :(

![photo_2024-03-19_13-49-38](https://github.com/organicmaps/organicmaps/assets/31618496/0b6f6a72-84d6-4d17-bcb5-8bcefe681c97)
![photo_2024-03-19_13-49-27](https://github.com/organicmaps/organicmaps/assets/31618496/c6627451-4b37-4088-9755-69841dfe0869)
![photo_2024-03-19_13-49-33](https://github.com/organicmaps/organicmaps/assets/31618496/eec386e5-abf8-4e72-a868-4d7d6fcd4ff4)

**Comment by pastk** [comment]
FYI @vng
